### PR TITLE
WIP - Try to fix compilation under Linux x64

### PR DESCRIPTION
**Pull request by berdal84**
I'm trying to compile my project Nodable under Windows and Linux and it appear that with the same CMAKE configuration both are not compiling. Window version well compiles, but the linux doesn't.

Here you can find two small try to fix some of the errors I had. 

Note that I had similar errors both on gcc and clang, the log show above is from clang under ubuntu.

I don't understand how to fix others errors relative to TypeDescGetter.

From a general point of view, I don't understand why it is not compiling by default on both Windows and Linux since you are not using any platform specific includes, am I right ?

```
berenger@Precision690:/mnt/e/Repositories/Nodable_dev$ cmake --build .
[  2%] Building CXX object CMakeFiles/Nodable.dir/sources/main.cpp.o
In file included from /mnt/e/Repositories/Nodable_dev/sources/main.cpp:8:
In file included from /mnt/e/Repositories/Nodable_dev/./sources/Core/Application.h:4:
In file included from /mnt/e/Repositories/Nodable_dev/./sources/Node/Node.h:6:
In file included from /mnt/e/Repositories/Nodable_dev/./sources/Core/Object.h:4:
In file included from /mnt/e/Repositories/Nodable_dev/extern/mirror/mirror.h:3:
/mnt/e/Repositories/Nodable_dev/extern/mirror/mirror_base.h:166:21: error: explicit specialization of 'mirror::TypeDescGetter<signed char, void, void>' after instantiation
        template <> struct TypeDescGetter<int8_t> { static TypeDesc* Get() { static TypeDesc s_typeDesc = TypeDesc(Type_int8); return &s_typeDesc; } };
                           ^~~~~~~~~~~~~~~~~~~~~~
/mnt/e/Repositories/Nodable_dev/extern/mirror/mirror_base.h:154:19: note: implicit instantiation first required here
                        case 1: return TypeDescGetter<int8_t>::Get();
                                       ^
/mnt/e/Repositories/Nodable_dev/extern/mirror/mirror_base.h:167:21: error: explicit specialization of 'mirror::TypeDescGetter<short, void, void>' after instantiation
        template <> struct TypeDescGetter<int16_t> { static TypeDesc* Get() { static TypeDesc s_typeDesc = TypeDesc(Type_int16); return &s_typeDesc; } };
                           ^~~~~~~~~~~~~~~~~~~~~~~
/mnt/e/Repositories/Nodable_dev/extern/mirror/mirror_base.h:155:19: note: implicit instantiation first required here
                        case 2: return TypeDescGetter<int16_t>::Get();
                                       ^
/mnt/e/Repositories/Nodable_dev/extern/mirror/mirror_base.h:168:21: error: explicit specialization of 'mirror::TypeDescGetter<int, void, void>' after instantiation
        template <> struct TypeDescGetter<int32_t> { static TypeDesc* Get() { static TypeDesc s_typeDesc = TypeDesc(Type_int32); return &s_typeDesc; } };
                           ^~~~~~~~~~~~~~~~~~~~~~~
/mnt/e/Repositories/Nodable_dev/extern/mirror/mirror_base.h:156:19: note: implicit instantiation first required here
                        case 4: return TypeDescGetter<int32_t>::Get();
                                       ^
/mnt/e/Repositories/Nodable_dev/extern/mirror/mirror_base.h:169:21: error: explicit specialization of 'mirror::TypeDescGetter<long, void, void>' after instantiation
        template <> struct TypeDescGetter<int64_t> { static TypeDesc* Get() { static TypeDesc s_typeDesc = TypeDesc(Type_int64); return &s_typeDesc; } };
                           ^~~~~~~~~~~~~~~~~~~~~~~
/mnt/e/Repositories/Nodable_dev/extern/mirror/mirror_base.h:157:19: note: implicit instantiation first required here
                        case 8: return TypeDescGetter<int64_t>::Get();
                                       ^
/mnt/e/Repositories/Nodable_dev/extern/mirror/mirror_base.h:207:4: error: static_assert failed "Mismatching pointer count between cast source and cast destination."
                        static_assert(false, "Mismatching pointer count between cast source and cast destination.");
                        ^             ~~~~~
/mnt/e/Repositories/Nodable_dev/extern/mirror/mirror_base.h:216:4: error: static_assert failed "Mismatching pointer count between cast source and cast destination."
                        static_assert(false, "Mismatching pointer count between cast source and cast destination.");
                        ^             ~~~~~
In file included from /mnt/e/Repositories/Nodable_dev/sources/main.cpp:8:
In file included from /mnt/e/Repositories/Nodable_dev/./sources/Core/Application.h:4:
In file included from /mnt/e/Repositories/Nodable_dev/./sources/Node/Node.h:6:
/mnt/e/Repositories/Nodable_dev/./sources/Core/Object.h:63:3: error: 'getClass' marked 'override' but does not override any member functions
                MIRROR_CLASS(Object)();
                ^
/mnt/e/Repositories/Nodable_dev/extern/mirror/mirror_macros.h:7:25: note: expanded from macro 'MIRROR_CLASS'
        virtual mirror::Class* getClass() const override { return _class::GetClass(); }\
                               ^
In file included from /mnt/e/Repositories/Nodable_dev/sources/main.cpp:8:
In file included from /mnt/e/Repositories/Nodable_dev/./sources/Core/Application.h:4:
/mnt/e/Repositories/Nodable_dev/./sources/Node/Node.h:44:15: error: no member named 'insert_or_assign' in 'std::map<std::__cxx11::basic_string<char>, Nodable::Component *, std::less<std::__cxx11::basic_string<char> >, std::allocator<std::pair<const std::__cxx11::basic_string<char>, Nodable::Component *> > >'
                        components.insert_or_assign(name, _component);
                        ~~~~~~~~~~ ^
/mnt/e/Repositories/Nodable_dev/./sources/Node/Node.h:84:25: error: use 'template' keyword to treat 'as' as a dependent template name
                                        return it->second->as<T>();
                                                           ^
                                                           template
/mnt/e/Repositories/Nodable_dev/./sources/Node/Node.h:91:18: error: member access into incomplete type 'Nodable::Component'
                                if (component->getClass()->isChildOf(c, false)) {
                                             ^
/mnt/e/Repositories/Nodable_dev/./sources/Core/Nodable.h:56:8: note: forward declaration of 'Nodable::Component'
        class Component;
              ^
In file included from /mnt/e/Repositories/Nodable_dev/sources/main.cpp:8:
In file included from /mnt/e/Repositories/Nodable_dev/./sources/Core/Application.h:4:
/mnt/e/Repositories/Nodable_dev/./sources/Node/Node.h:92:22: error: member access into incomplete type 'Nodable::Component'
                                        return component->as<T>();
                                                        ^
/mnt/e/Repositories/Nodable_dev/./sources/Core/Nodable.h:56:8: note: forward declaration of 'Nodable::Component'
        class Component;
              ^
In file included from /mnt/e/Repositories/Nodable_dev/sources/main.cpp:8:
In file included from /mnt/e/Repositories/Nodable_dev/./sources/Core/Application.h:5:
In file included from /mnt/e/Repositories/Nodable_dev/./sources/Component/History.h:14:
/mnt/e/Repositories/Nodable_dev/./sources/Component/Container.h:18:30: warning: 'update' overrides a member function but is not marked 'override' [-Winconsistent-missing-override]
                bool                            update();
                                                ^
/mnt/e/Repositories/Nodable_dev/./sources/Node/Node.h:111:23: note: overridden virtual function is here
                virtual bool        update            ();
                                    ^
In file included from /mnt/e/Repositories/Nodable_dev/sources/main.cpp:8:
In file included from /mnt/e/Repositories/Nodable_dev/./sources/Core/Application.h:6:
/mnt/e/Repositories/Nodable_dev/./sources/Core/File.h:28:36: warning: 'update' overrides a member function but is not marked 'override' [-Winconsistent-missing-override]
                bool                             update();
                                                 ^
/mnt/e/Repositories/Nodable_dev/./sources/Node/Node.h:111:23: note: overridden virtual function is here
                virtual bool        update            ();
                                    ^
In file included from /mnt/e/Repositories/Nodable_dev/sources/main.cpp:8:
/mnt/e/Repositories/Nodable_dev/./sources/Core/Application.h:31:20: warning: 'update' overrides a member function but is not marked 'override' [-Winconsistent-missing-override]
                bool             update();
                                 ^
/mnt/e/Repositories/Nodable_dev/./sources/Node/Node.h:111:23: note: overridden virtual function is here
                virtual bool        update            ();
                                    ^
In file included from /mnt/e/Repositories/Nodable_dev/sources/main.cpp:9:
In file included from /mnt/e/Repositories/Nodable_dev/./sources/Core/Test.h:13:
/mnt/e/Repositories/Nodable_dev/./sources/Node/Variable.h:29:30: error: 'T' does not refer to a value
                        return  get("value")->get<T>();
                                                  ^
/mnt/e/Repositories/Nodable_dev/./sources/Node/Variable.h:27:22: note: declared here
                template <typename T>
                                   ^
/mnt/e/Repositories/Nodable_dev/./sources/Node/Variable.h:29:26: error: no member named 'get' in 'Nodable::Member'; did you mean 'set'?
                        return  get("value")->get<T>();
                                              ^~~
                                              set
/mnt/e/Repositories/Nodable_dev/./sources/Core/Member.h:27:23: note: 'set' declared here
                void                set(const Member*);
                                    ^
In file included from /mnt/e/Repositories/Nodable_dev/sources/main.cpp:9:
In file included from /mnt/e/Repositories/Nodable_dev/./sources/Core/Test.h:13:
/mnt/e/Repositories/Nodable_dev/./sources/Node/Variable.h:29:33: error: expected expression
                        return  get("value")->get<T>();
                                                     ^
3 warnings and 14 errors generated.
make[2]: *** [CMakeFiles/Nodable.dir/build.make:63: CMakeFiles/Nodable.dir/sources/main.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:76: CMakeFiles/Nodable.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```

### CODE REVIEW

**Comment by berdal84** [comment]
I finally found that I had to move the declaration of the implicit templates before the explicit (for TypeDescGetter)

File: mirror_macros.h [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 #define MIRROR_CLASS(_class, ...)\
 public:\
-	virtual mirror::Class* getClass() const { return _class::GetClass(); }\
```
Comment (Grouflon) [inline mirror_macros.h:7]: won't this cause issue compilation on the topmost parent ? does this works if the class is not actually overriding anything ?

File: mirror_macros.h [hunk 2]
```diff
@@ -4,7 +4,7 @@
...
+	virtual mirror::Class* getClass() const override { return _class::GetClass(); }\
```
Comment (Grouflon) [inline mirror_macros.h:7]: won't this cause issue compilation on the topmost parent ? does this works if the class is not actually overriding anything ?

Comment (berdal84) [reply mirror_macros.h:7]: You're right. I have an error for my base class Object. I exchanged N warnings for a single error ^^

**Comment by berdal84** [comment]
The main problems in compiling is that templates in templates are not compiling well. 
I think this is due to the implit template instantiation policy in the compiler :
https://stackoverflow.com/questions/40603219/when-does-implicit-template-instantiation-occur

**Comment by berdal84** [comment]
There is an option on gcc called `-fno-implicit-templates`:
http://www-tcad.stanford.edu/local/DOC/g++FAQ_41.html

File: mirror_base.h [hunk 3]
```diff
@@ -176,6 +161,23 @@ namespace mirror
...
+			case 1: return TypeDescGetter<int8_t>::Get();
+			case 2: return TypeDescGetter<int16_t>::Get();
+			case 4: return TypeDescGetter<int32_t>::Get();
+			case 8: return TypeDescGetter<int64_t>::Get();
+			}
+			return nullptr;
+		}
+	};
+
+
```
Comment (Grouflon) [inline mirror_base.h:180]: I dont really get why this had to move below GetTypeDesc, can you elaborate please ? is it still working as intended with that modification ?

File: mirror_base.h [hunk 4]
```diff
@@ -139,23 +139,8 @@ namespace mirror
...
-	};
-
-	template <typename T>
-	struct TypeDescGetter<T, void, std::enable_if_t<std::is_enum<T>::value>>
-	{
-		static TypeDesc* Get()
-		{
-			switch (sizeof(T))
-			{
-			case 1: return TypeDescGetter<int8_t>::Get();
```
Comment (berdal84) [inline mirror_base.h:153]: Previously, here the `typeDescGetter<int8_t>` do not exists yet. By moving the declaration of `struct TypeDescGetter<T, void, std::enable_if_t<std::is_enum<T>::value>>` we are sure all TypeDescGetter<T> will exists (for in8_t, 16, 32 and 64)

File: mirror_base.h (hunk 3 above)
Comment (berdal84) [reply mirror_base.h:180]: I just added a comment for the move reason just above. I don't know if it works, but the related errors disappear.

**Review by berdal84 — COMMENTED** [review]
You can close the merge request when you read my answers. 
For two reasons:
- The nested templates compilation problem occurs with the Nodable code base too, not just on mirror.
- There must be a compilation option to avoid instantiation of implicit templates before the explicit one.

Still unsolved problem to me (relative to cmake):
- Why using CMake my code do not compile well on both windows and linux... ?

**Pull request closed without merging**